compiler/rustc_lint/src/non_local_def.rs

```diff
@@ -207,6 +266,47 @@ impl<'tcx> LateLintPass<'tcx> for NonLocalDefinitions {
     }
 }
 
+/// Replace every local type by inference variable.
+///
+/// ```text
+/// <Global<Local> as std::cmp::PartialEq<Global<Local>>>
+/// to
+/// <Global<_> as std::cmp::PartialEq<Global<_>>>
+/// ```
+struct LocalTypeInferenceFolder<'a, 'tcx> {
+    infcx: &'a InferCtxt<'tcx>,
+    to_ignore: u32,
+    impl_parent: DefId,
+    impl_parent_parent: Option<DefId>,
+    span: Span,
+}
+
+impl<'a, 'tcx> TypeFolder<TyCtxt<'tcx>> for LocalTypeInferenceFolder<'a, 'tcx> {
+    fn interner(&self) -> TyCtxt<'tcx> {
+        self.infcx.tcx
+    }
+
+    fn fold_ty(&mut self, t: Ty<'tcx>) -> Ty<'tcx> {
+        if let Some(ty_did) = t.ty_def_id()
+            && did_has_local_parent(
+                ty_did,
+                self.infcx.tcx,
+                self.impl_parent,
+                self.impl_parent_parent,
+            )
+            && self.to_ignore == 0
```

Comment:
why this? to allow any impl whose self type is a local type to get skipped. Please instead specialcase such impls directly. Using an undocumented integer `to_ignore` for feels confusing

Reply:
I added this to be able to skip the first "type" if it's a local type.

So `<Local<Local> as std::cmp::PartialEq<Global<Local>>>` becomes `<Local<_> as std::cmp::PartialEq<Global<_>>>` (and not `<_ as std::cmp::PartialEq<Global<_>>>`). (Is this even right?)

I did it this way with a counter/bool because I don't know how to skip the first type while still recursing trough it. If you know a better way of doing it, I'm all for it.

Reply:
Impls whose self-type is a local type are always local. You can check that explicitly first and in that case completely avoid the new check.

Reply:
Let's do that. (this might also save some unnecessary work, which is always good to take)